zmq-async/src/zmq_async.ml

```diff
@@ -0,0 +1,91 @@
+open Base
+open Async_kernel
+open Async_unix
+
+module Socket = struct
+  exception Break_event_loop [@@deriving sexp_of]
+  exception Retry [@@deriving sexp_of]
+
+  type 'a t =
+    { socket : 'a ZMQ.Socket.t sexp_opaque
+    ; fd : Fd.t
+    }
+  [@@deriving sexp_of]
+
+  let to_socket t = t.socket
+
+  let of_socket socket =
+    let fd =
+      Fd.create
+        (Fd.Kind.Socket `Active)
+        (ZMQ.Socket.get_fd socket)
+        (Info.of_string "<zmq>")
+    in
+    { socket; fd }
+
+  let zmq_event socket ~f =
+    let open ZMQ.Socket in
+    try begin match events socket with
```

Comment:
I think it would be nicer to pull the exception matching inside the `match` and get rid of the `try` altogether, which also lets one to avoid `begin`/`end`.